### ISSUE TITLE
`fillSlot` should guard against not allowed state transitions

### ISSUE THREAD

**Issue by AuHau** [codex-storage/codex-contracts-eth#157]
`fillSlot` should verify that the Request is in states `New` or `Started`, if we would allow for the rest of the states then we could get illegal state transitions from `Cancelled`/`Failed`/`Finished` to `Started`.

Credit to @0x-r4bbit as he found this as part of Certora work.

**Comment by 0x-r4bbit** [comment]
Working on that atm.

**Comment by 0x-r4bbit** [comment]
Quick update on this, we have a fix for this in place but yet have to fix some rules that are failing after we've landed #158

**Comment by 0x-r4bbit** [comment]
This will be fixed with #161

**Comment by AuHau** [comment]
In the end it seems like this line actually prevent the behavior described in this issue:
https://github.com/codex-storage/codex-contracts-eth/blob/8bbb4257f7b42bf257303d4f504674adbebdefee/contracts/Marketplace.sol#L139

**Comment by 0x-r4bbit** [comment]
Closing this as discussed above.

**Comment by 0x-r4bbit** [comment]
Just to add this here for reference:

The reason this issue came up in the first place, is because we had a counter example in one of the rules that has shown that an already `Finished` request can be changed to `Started`, which is actually forbidden.

The reason it could do that was because the prover starts with a finished request and calls `fillSlot()` with a slotId that points to a request with id `0`. In practice, id `0` is not really possible, but the prover doesn't know that.

When `requestId == 0` then this ultimately causes the code to think the `slotstate` of the slot in question is `Free` (which is not necessarily the case).

Long story short, the prover shouldn't call the function with `requestId = 0` but we also can't blindly require that the `requestId != 0` because there's other cases that would be excluded from the rule.

For more context see: https://github.com/codex-storage/codex-contracts-eth/pull/162#discussion_r1728772997
